### DISCUSSION TITLE
RFC: Preview Mode

### DISCUSSION THREAD

**Discussion by migueloller** [makeswift/makeswift#142] [Ideas]
# Background

When integrating your Next.js app into Makeswift you have to create a preview page (e.g., `pages/makeswift.js`). This page is is rendered by Makeswift in an `iframe` and communicates with the Makeswift builder via `postMessage` so you can visually edit your Next.js app.

When editing in the Makeswift builder you expect to see your page’s latest draft, not the published version. That’s why the preview page uses `getServerSideProps` and requests the page’s draft while other pages most likely use `getStaticProps` and request the published version of the page.

Having this separate preview page has worked so far, but there are a couple of problems with it:

- Increases the surface area of the Makeswift integration and doesn’t put developers in the [pit of success](https://blog.codinghorror.com/falling-into-the-pit-of-success/)
    - Requires creating a preview page that wouldn’t exist in your Next.js app otherwise
    - Requires providing the route’s path when adding the host to the Makeswift site settings, which is prone to errors and hard for us to automatically fix for you
    - Creates a fork between what you see in the builder and your live pages, which goes against Makeswift’s principle of honoring a *true* [WYSIWYG](https://en.wikipedia.org/wiki/WYSIWYG) experience
- Hurts incremental adoption since there is only one preview page but multiple other pages in your Next.js app, which results in only 1 type page component being editable in the builder
- When Makeswift wants to show you a preview of your site in its draft state, you can’t just navigate around your site (i.e., by clicking links) since that would navigate you away from the single preview page

# Proposal

We propose adding native support for Next.js’ [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode) to Makeswift.

When using the Makeswift builder (i.e., editing the draft version of the page in the `iframe`) Preview Mode should *always* be **on**.

When visiting a published page, Preview Mode should *always* be **off**.

The goal is to make it transparent both for developers integrating Makeswift with their Next.js app and end-users using the Makeswift app to do visually edits.

Read on for details about how this will be implemented, what we’ve explored, and future plans ☕

## Makeswift API Route

Inspired by [NextAuth.js](https://next-auth.js.org/) we propose an API route that must be created when integrating Makeswift into your Next.js app: `pages/api/makeswift/[...makeswift].js`.

This API route will be used to handle Preview Mode functionality:

- Preview Mode can be enabled at `pages/api/makeswift/preview`
- Preview Mode can be disabled at `pages/api/makeswift/clear-preview-mode-cookies`

<details>
<summary>

### Future Plans

</summary>

In addition to handling Preview Mode, the Makeswift API route will also be used for other features we’re currently working on:

- Leveraging [On-demand Revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta) so that your pages can be statically generated and automatically revalidated when changing a Makeswift color, typography, component, or other resources.
- Providing a manifest of *all* the pages in your Next.js app to the Makeswift UI—even if you haven’t integrated them yet—allowing your teammates to get a full view of the app as you incrementally adopt Makeswift.

</details>

## Toggling Preview Mode On & Off

After trying [different approaches](#alternative-approaches), we found that the most reliable way to ensure Preview Mode is always-on in the Makeswift builder `iframe` and always-off when not in the `iframe` was to use JavaScript to detect the `iframe` and then redirecting to the relevant API route that configures Preview Mode.

```jsx
import Router from 'next/router'

const isInIframe = window.parent !== window
const isPreview = Router.isPreview

if (isInIframe && !isPreview) {
  const returnTo = window.location.href
  const urlSearchParams = new URLSearchParams({ returnTo })

  window.location.replace(`/api/makeswift/preview?${urlSearchParams.toString()}`)
} else if (!isInIframe && isPreview) {
  const returnTo = window.location.href
  const urlSearchParams = new URLSearchParams({ returnTo })

  window.location.href = `/api/makeswift/clear-preview-mode-cookies?${urlSearchParams.toString()}`
}
```

### Configuring Preview Mode on initial page load with a `script` in the `head`

To avoid a flash of incorrect data we need to make sure that our logic runs as soon as possible. We could try using `next/head` but placing a `script` in `next/head` results in a [Next.js warning](https://nextjs.org/docs/messages/no-script-tags-in-head-component) to use `next/script` instead because of potential issues with Suspense and/or streaming.

`next/script` could be used with a strategy of `beforeInteractive` but this is only allowed in `_document.js`. We can check for Preview Mode in `_document.js` using `props.__NEXT_DATA__.isPreview`. Adding this logic to `_document.js` would make it harder for us to eventually remove `_document.js` as part of integrating Makeswift to your Next.js app.

<aside>

🚧 We will continue to explore more solutions and collaborate with the Next.js team to make sure this works well with Suspense and streaming once those features reach GA.

</aside>

### Configuring Preview Mode on client-side navigation by patching `fetch` and `XMLHttpRequest`

When doing client-side route changes, things get a bit trickier. Next.js will usually render the next page with the current Preview Mode state before the browser navigates away. The most reliable approach that we found was patching `fetch` and `XMLHttpRequest` so that we can ensure the correct values for Preview Mode cookies are set. We’ve had success monkey-patching the Next.js router’s `getRouteInfo` method, though, doing something like this:

```jsx
let patched = false

function patch() {
  if (patched) return

  patched = true

  const initialPreviewModeCookies = getPreviewModeCookies()
  const originalFetch = window.fetch

  window.fetch = function patchedFetch(...args) {
    persistInitialPreviewModeCookies(initialPreviewModeCookies)

    return originalFetch.apply(this, args)
  }

  const originalSend = XMLHttpRequest.prototype.send

  XMLHttpRequest.prototype.send = function patchedSend(...args) {
    persistInitialPreviewModeCookies(initialPreviewModeCookies)

    return originalSend.apply(this, args)
  }
}
```

# Backwards Compatibility

We will be deprecating the preview page in favor of the new Makeswift API route and Next.js Preview Mode.

# Alternative Approaches

## Rewriting to an API route to proxy the page and inject Preview Mode cookies

One approach we considered early on was creating a rewrite rule to an API route that proxies all pages and configures Preview Mode. The rewrites could be added automatically by a Makeswift Next.js plugin.

```jsx

async rewrites() {
  return {
    beforeFiles: [
      {
        source: '/:path*',
        has: [{ type: 'query', key: 'makeswift_preview_mode', value: 'true' }],
        destination: '/api/makeswift/preview?path=:path*',
      },
    ],
  }
}
```

This approach is not only the simplest in terms of integration surface area, but also potentially the most reliable, since it would handle both initial load and subsequent client-side navigation since Next.js’ router applies rewrites in the server and client.

Unfortunately this doesn’t work because during hydration, Next.js treats the rewrite destination as a page and attempts to load it at `_next/static/chunks/pages/api/makeswift/preview.js`, which results in a 404 error. This error then causes a reload of the page, and once hydration happens, the request fails again. This goes on and on in an infinite loop.

## Patching Next.js’ router instead of `fetch` and `XMLHttpRequest`

Before deciding to patch `fetch` and `XMLHttpRequest` we explored patching Next.js’ router but found that this would be unreliable and would break as internal Next.js APIs changed. Patching platform APIs is much more reliable and stable.

**Comment by Galaxspheria** [comment]
Looks great! One question - would this mean that preview mode is enabled if the site is loaded in any iframe, not just the Makeswift preview one? What if I'd like to be able to embed my site elsewhere without enabling these preview features?

Could be worth exploring a quick back-and-forth with `window.postMessage()` so that Makeswift can inform the site within the iframe that it is indeed being hosted within a Makeswift preview, not just any iframe.

**Reply by migueloller** [reply]
> Looks great! One question - would this mean that preview mode is enabled if the site is loaded in any iframe, not just the Makeswift preview one? What if I'd like to be able to embed my site elsewhere without enabling these preview features?

@Galaxspheria, great question! The answer to that is that it will only happen in the Makeswift builder since only the Makeswift builder will have the authenticated user's secret, provided via a URL search param, which is needed for Preview Mode to turn on. If you embed the site in another iframe it will behave normally.

---

> Could be worth exploring a quick back-and-forth with window.postMessage() so that Makeswift can inform the site within the iframe that it is indeed being hosted within a Makeswift preview, not just any iframe.

What makes the integration work "magically" is that Preview Mode is enforced in a `<head>` script to guarantee there's not a flash of preview/non-preview content. JavaScript in the `<head>` element is "blocking" and runs before paint. If we did something asynchronous, like `postMessage` then you might see a flash of the preview page outside the iframe or a flash of the live page in the builder iframe.

**Reply by Galaxspheria** [reply]
Ah makes sense, even better! Slick that you can do all that validation before first paint.

**Reply by migueloller** [reply]
Yeah, it was a very cool idea/feature/problem to solve for in a way that worked reliably while having a very minimal API surface area. Here's some code if you wanna check it out: https://github.com/makeswift/makeswift/blob/f2c8198cb19fca2654756736556280d42db0c985/packages/runtime/src/next/preview-mode.tsx

**Comment by migueloller** [comment]
Preview Mode support has been released in v0.2.0!
